Oracle Invalid column type: 16 when Binding null Boolean values at insert

### Expected Behavior

Binding null NUMBER columns to Boolean type works as expected. Same expected at insert.

### Actual Behaviour

"Invalid column type: 16" exception is thrown as below

I have found the issue is this line: https://github.com/micronaut-projects/micronaut-data/blob/a699b42051770e326a742fda5e443c4b47de71d6/data-jdbc/src/main/java/io/micronaut/data/jdbc/mapper/JdbcQueryStatement.java#L77

If I change Types.BOOLEAN to Types.NUMBERIC or Types.INTEGER it works.

```
io.micronaut.data.exceptions.DataAccessException: SQL Error executing INSERT: Error setting JDBC null value: Invalid column type: 16
	at io.micronaut.data.runtime.operations.internal.AbstractRepositoryOperations.persistOne(AbstractRepositoryOperations.java:289)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.lambda$persist$16(DefaultJdbcRepositoryOperations.java:750)
	at io.micronaut.transaction.support.AbstractSynchronousTransactionManager.executeWrite(AbstractSynchronousTransactionManager.java:175)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.persist(DefaultJdbcRepositoryOperations.java:748)
	at io.micronaut.data.runtime.intercept.DefaultSaveEntityInterceptor.intercept(DefaultSaveEntityInterceptor.java:45)
	at io.micronaut.data.intercept.DataIntroductionAdvice.intercept(DataIntroductionAdvice.java:114)
	at io.micronaut.data.intercept.DataIntroductionAdvice.intercept(DataIntroductionAdvice.java:88)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:137)
	at io.micronaut.validation.ValidatingInterceptor.intercept(ValidatingInterceptor.java:138)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:137)
	at com.example.dao.SampleRepository$Intercepted.save(Unknown Source)
	at com.example.controller.SampleController.createSample(SampleController.java:25)
	at com.example.controller.$SampleController$Definition$Exec.dispatch(Unknown Source)
	at io.micronaut.context.AbstractExecutableMethodsDefinition$DispatchedExecutableMethod.invoke(AbstractExecutableMethodsDefinition.java:351)
	at io.micronaut.context.DefaultBeanContext$4.invoke(DefaultBeanContext.java:584)
	at io.micronaut.web.router.AbstractRouteMatch.execute(AbstractRouteMatch.java:303)
	at io.micronaut.web.router.RouteMatch.execute(RouteMatch.java:111)
	at io.micronaut.http.context.ServerRequestContext.with(ServerRequestContext.java:103)
	at io.micronaut.http.server.RouteExecutor.lambda$executeRoute$14(RouteExecutor.java:649)
	at reactor.core.publisher.FluxDeferContextual.subscribe(FluxDeferContextual.java:49)
	at reactor.core.publisher.Flux.subscribe(Flux.java:8402)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:426)
	at io.micronaut.reactive.reactor.instrument.ReactorSubscriber.onNext(ReactorSubscriber.java:57)
	at reactor.core.publisher.MonoCreate$DefaultMonoSink.success(MonoCreate.java:160)
	at io.micronaut.http.server.netty.RoutingInBoundHandler$4.doOnComplete(RoutingInBoundHandler.java:880)
	at io.micronaut.core.async.subscriber.CompletionAwareSubscriber.onComplete(CompletionAwareSubscriber.java:71)
	at io.micronaut.http.server.netty.jackson.JsonContentProcessor$1.doOnComplete(JsonContentProcessor.java:136)
	at io.micronaut.core.async.subscriber.CompletionAwareSubscriber.onComplete(CompletionAwareSubscriber.java:71)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.micronaut.core.async.processor.SingleThreadedBufferingProcessor.doOnComplete(SingleThreadedBufferingProcessor.java:48)
	at io.micronaut.jackson.core.parser.JacksonCoreProcessor.doOnComplete(JacksonCoreProcessor.java:94)
	at io.micronaut.core.async.subscriber.SingleThreadedBufferingSubscriber.onComplete(SingleThreadedBufferingSubscriber.java:71)
	at io.micronaut.http.server.netty.jackson.JsonContentProcessor.doOnComplete(JsonContentProcessor.java:161)
	at io.micronaut.core.async.subscriber.CompletionAwareSubscriber.onComplete(CompletionAwareSubscriber.java:71)
	at io.micronaut.http.netty.reactive.HandlerPublisher.complete(HandlerPublisher.java:416)
	at io.micronaut.http.netty.reactive.HandlerPublisher.handlerRemoved(HandlerPublisher.java:403)
	at io.netty.channel.AbstractChannelHandlerContext.callHandlerRemoved(AbstractChannelHandlerContext.java:946)
	at io.netty.channel.DefaultChannelPipeline.callHandlerRemoved0(DefaultChannelPipeline.java:637)
	at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:477)
	at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:423)
	at io.micronaut.http.netty.stream.HttpStreamsHandler.removeHandlerIfActive(HttpStreamsHandler.java:465)
	at io.micronaut.http.netty.stream.HttpStreamsHandler.handleReadHttpContent(HttpStreamsHandler.java:312)
	at io.micronaut.http.netty.stream.HttpStreamsHandler.channelRead(HttpStreamsHandler.java:275)
	at io.micronaut.http.netty.stream.HttpStreamsServerHandler.channelRead(HttpStreamsServerHandler.java:123)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at io.netty.handler.codec.http.HttpServerKeepAliveHandler.channelRead(HttpServerKeepAliveHandler.java:64)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.flow.FlowControlHandler.dequeue(FlowControlHandler.java:200)
	at io.netty.handler.flow.FlowControlHandler.read(FlowControlHandler.java:139)
	at io.netty.channel.AbstractChannelHandlerContext.invokeRead(AbstractChannelHandlerContext.java:686)
	at io.netty.channel.AbstractChannelHandlerContext.read(AbstractChannelHandlerContext.java:671)
	at io.micronaut.http.netty.reactive.HandlerPublisher.requestDemand(HandlerPublisher.java:163)
	at io.micronaut.http.netty.stream.HttpStreamsHandler$2.requestDemand(HttpStreamsHandler.java:266)
	at io.micronaut.http.netty.reactive.HandlerPublisher$ChannelSubscription.receivedDemand(HandlerPublisher.java:547)
	at io.micronaut.http.netty.reactive.HandlerPublisher$ChannelSubscription.lambda$request$0(HandlerPublisher.java:474)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute$$$capture(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: io.micronaut.data.exceptions.DataAccessException: Error setting JDBC null value: Invalid column type: 16
	at io.micronaut.data.jdbc.mapper.JdbcQueryStatement.setDynamic(JdbcQueryStatement.java:119)
	at io.micronaut.data.jdbc.mapper.JdbcQueryStatement.setDynamic(JdbcQueryStatement.java:35)
	at io.micronaut.data.runtime.operations.internal.AbstractSqlRepositoryOperations.setStatementParameter(AbstractSqlRepositoryOperations.java:215)
	at io.micronaut.data.runtime.operations.internal.StoredSqlOperation.setStatementParameter(StoredSqlOperation.java:244)
	at io.micronaut.data.runtime.operations.internal.StoredSqlOperation.setParameters(StoredSqlOperation.java:223)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations$JdbcEntityOperations.setParameters(DefaultJdbcRepositoryOperations.java:1019)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations$JdbcEntityOperations.setParameters(DefaultJdbcRepositoryOperations.java:978)
	at io.micronaut.data.runtime.operations.internal.AbstractRepositoryOperations.lambda$persistOne$1(AbstractRepositoryOperations.java:279)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.prepareStatement(DefaultJdbcRepositoryOperations.java:368)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.prepareStatement(DefaultJdbcRepositoryOperations.java:129)
	at io.micronaut.data.runtime.operations.internal.AbstractRepositoryOperations.persistOne(AbstractRepositoryOperations.java:278)
	... 81 common frames omitted
Caused by: java.sql.SQLException: Invalid column type: 16
	at oracle.jdbc.driver.OracleStatement.getInternalType(OracleStatement.java:4469)
	at oracle.jdbc.driver.OraclePreparedStatement.setNullCritical(OraclePreparedStatement.java:4382)
	at oracle.jdbc.driver.OraclePreparedStatement.setNull(OraclePreparedStatement.java:4366)
	at oracle.jdbc.driver.OraclePreparedStatementWrapper.setNull(OraclePreparedStatementWrapper.java:937)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.setNull(HikariProxyPreparedStatement.java)
	at io.micronaut.data.jdbc.mapper.JdbcQueryStatement.setDynamic(JdbcQueryStatement.java:77)
	... 91 common frames omitted
```

### Steps To Reproduce

CrudRepository.save with a null Boolean entity is enough to reproduce.



### Environment Information

- Arch Linux
- Java 11
- Tested both the Git version and the maven (3.2.0) version

### Example Application

_No response_

### Version

3.2.3-SNAPSHOT

--- a/data-jdbc/src/main/java/io/micronaut/data/jdbc/mapper/JdbcQueryStatement.java
+++ b/data-jdbc/src/main/java/io/micronaut/data/jdbc/mapper/JdbcQueryStatement.java
@@ -19,6 +19,8 @@ import io.micronaut.core.annotation.Internal;
 import io.micronaut.core.annotation.NonNull;
 import io.micronaut.core.convert.ConversionService;
 import io.micronaut.data.exceptions.DataAccessException;
+import io.micronaut.data.jdbc.config.DataJdbcConfiguration;
+import io.micronaut.data.model.query.builder.sql.Dialect;
 import io.micronaut.data.runtime.convert.DataConversionService;
 import io.micronaut.data.runtime.mapper.QueryStatement;
 import io.micronaut.data.model.DataType;
@@ -37,6 +39,7 @@ import java.util.Date;
 public class JdbcQueryStatement implements QueryStatement<PreparedStatement, Integer> {
 
     private final ConversionService conversionService;
+    private final DataJdbcConfiguration jdbcConfiguration;
 
     public JdbcQueryStatement() {
         this(null);
@@ -49,22 +52,43 @@ public class JdbcQueryStatement implements QueryStatement<PreparedStatement, Int
      * @since 3.1
      */
     public JdbcQueryStatement(DataConversionService conversionService) {
-        // Backwards compatibility should be removed in the next version
-        this.conversionService = conversionService == null ? ConversionService.SHARED : conversionService;
+        this(conversionService, new DataJdbcConfiguration("default"));
+    }
+
+    /**
+     * Constructs a new instance.
+     *
+     * @param conversionService The data conversion service
+     * @param jdbcConfiguration The JDBC configuration
+     * @since 4.6.1
+     */
+    public JdbcQueryStatement(DataConversionService conversionService, DataJdbcConfiguration jdbcConfiguration) {
+        this.conversionService = conversionService;
+        this.jdbcConfiguration = jdbcConfiguration;
     }
 
     /**
      * Find the SQL type from {@link DataType}.
+     *
      * @param dataType The data type
+     * @param dialect The dialect
      * @return The SQL type
      */
     @Internal
-    public static int findSqlType(@NonNull DataType dataType) {
+    public static int findSqlType(@NonNull DataType dataType, @NonNull Dialect dialect) {
         return switch (dataType) {
             case LONG -> Types.BIGINT;
             case STRING, JSON -> Types.VARCHAR;
             case DATE -> Types.DATE;
-            case BOOLEAN -> Types.BOOLEAN;
+            case BOOLEAN -> {
+                if (dialect == Dialect.ORACLE) {
+                    // oracle driver treats Boolean types as bits
+                    // see https://github.com/micronaut-projects/micronaut-data/issues/1259
+                    yield Types.BIT;
+                } else {
+                    yield Types.BOOLEAN;
+                }
+            }
             case INTEGER -> Types.INTEGER;
             case TIMESTAMP -> Types.TIMESTAMP;
             case TIME -> Types.TIME;
@@ -96,7 +120,7 @@ public class JdbcQueryStatement implements QueryStatement<PreparedStatement, Int
                         return this;
                     }
                     default -> {
-                        int sqlType = findSqlType(dataType);
+                        int sqlType = findSqlType(dataType, jdbcConfiguration.getDialect());
                         if (sqlType != -1) {
                             statement.setNull(index, sqlType);
                         } else if (dataType.isArray()) {

--- a/data-jdbc/src/main/java/io/micronaut/data/jdbc/operations/DefaultJdbcRepositoryOperations.java
+++ b/data-jdbc/src/main/java/io/micronaut/data/jdbc/operations/DefaultJdbcRepositoryOperations.java
@@ -199,7 +199,7 @@ public final class DefaultJdbcRepositoryOperations extends AbstractSqlRepository
             dataSourceName,
             new ColumnNameResultSetReader(conversionService),
             new ColumnIndexResultSetReader(conversionService),
-            new JdbcQueryStatement(conversionService),
+            new JdbcQueryStatement(conversionService, jdbcConfiguration),
             dateTimeProvider,
             entityRegistry,
             beanContext,
@@ -565,7 +565,7 @@ public final class DefaultJdbcRepositoryOperations extends AbstractSqlRepository
             preparedQuery.bindParameters(new JdbcParameterBinder(connection, callableStatement, preparedQuery));
             if (!preparedQuery.getResultArgument().isVoid()) {
                 DataType resultDataType = preparedQuery.getResultDataType();
-                int sqlType = JdbcQueryStatement.findSqlType(resultDataType);
+                int sqlType = JdbcQueryStatement.findSqlType(resultDataType, jdbcConfiguration.getDialect());
                 int outIndex = preparedQuery.getQueryBindings().size() + 1;
                 callableStatement.registerOutParameter(outIndex, sqlType);
             }

--- a/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/oraclexe/bool/BooleanTest.java
+++ b/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/oraclexe/bool/BooleanTest.java
@@ -1,0 +1,17 @@
+package io.micronaut.data.jdbc.oraclexe.bool;
+
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.data.annotation.GeneratedValue;
+import io.micronaut.data.annotation.Id;
+import io.micronaut.data.annotation.MappedEntity;
+import jakarta.persistence.Column;
+
+@MappedEntity
+public record BooleanTest(
+    @Id @GeneratedValue
+    Long id,
+    @Column(nullable = true)
+    @Nullable
+    Boolean canBeNull
+) {
+}

--- a/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/oraclexe/bool/OracleBooleanTestRepository.java
+++ b/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/oraclexe/bool/OracleBooleanTestRepository.java
@@ -1,0 +1,9 @@
+package io.micronaut.data.jdbc.oraclexe.bool;
+
+import io.micronaut.data.jdbc.annotation.JdbcRepository;
+import io.micronaut.data.model.query.builder.sql.Dialect;
+import io.micronaut.data.repository.CrudRepository;
+
+@JdbcRepository(dialect = Dialect.ORACLE)
+public interface OracleBooleanTestRepository extends CrudRepository<BooleanTest, Long> {
+}

--- a/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/oraclexe/bool/OracleXEBooleanTypeSpec.groovy
+++ b/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/oraclexe/bool/OracleXEBooleanTypeSpec.groovy
@@ -1,0 +1,43 @@
+package io.micronaut.data.jdbc.oraclexe.bool
+
+import io.micronaut.context.ApplicationContext
+import io.micronaut.data.jdbc.oraclexe.OracleTestPropertyProvider
+import spock.lang.AutoCleanup
+import spock.lang.Shared
+import spock.lang.Specification
+
+class OracleXEBooleanTypeSpec extends Specification implements OracleTestPropertyProvider {
+    @AutoCleanup
+    @Shared
+    ApplicationContext applicationContext = ApplicationContext.run(properties)
+
+    void "test persist oracle boolean type as null"() {
+        given:
+        def repository = applicationContext.getBean(OracleBooleanTestRepository)
+
+        when:
+        def result = repository.save(new BooleanTest(null, null))
+        result = repository.findById(result.id()).orElse(null)
+
+        then:
+        result != null
+        result.canBeNull() == null
+
+        when:
+        repository.update(new BooleanTest(result.id(), true))
+        result = repository.findById(result.id()).orElse(null)
+
+        then:
+        result != null
+        result.canBeNull()
+
+        when:
+        repository.update(new BooleanTest(result.id(), null))
+        result = repository.findById(result.id()).orElse(null)
+
+        then:
+        result != null
+        result.canBeNull() == null
+
+    }
+}

